sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -140,6 +140,15 @@ trait PredicateHelper extends AliasHelper with Logging {
         findExpressionAndTrackLineageDown(replaceAlias(exp, aliasMap), a.child)
       case l: LeafNode if exp.references.subsetOf(l.outputSet) =>
         Some((exp, l))
+      case u: Union =>
+        val index = u.output.indexWhere(_.semanticEquals(exp))
+        if (index > -1) {
+          u.children
+            .flatMap(child => findExpressionAndTrackLineageDown(child.output(index), child))
+            .sortBy(_._2.stats.sizeInBytes).lastOption
```

Comment:
can you explain a bit more about this sorting part?

Reply:
In general, tables with larger sizes are more likely to be partition tables.

Reply:
and we only add DPP filter in one child of Union?

Reply:
For all child of Union expect the references is empty. For example: https://github.com/apache/spark/pull/36724

Reply:
shall we do the same for `case other =>`?

Reply:
After reading the code more, I don't think we have to use a heuristic here. We can let caller side to pass a callback function so that we can go to all the branches to find partitioned tables.

Reply:
OK